### PR TITLE
[automate-1882] Optionally add project teams if checkbox selected

### PR DESCRIPTION
**Pull request by tylercloke**
### :nut_and_bolt: Description: What code changed, and why?

Optionally add project teams if checkbox selected.

We should be good on length violation concerns since teams have super non-restrictive IDs atm:

https://github.com/chef/automate/blob/master/api/interservice/teams/teams.proto#L16-L20

### :white_check_mark: Checklist

- [ ] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [ ] Tests added/updated?
- [ ] Docs added/updated?
- [ ] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### :camera: Screenshots, if applicable

If one of the team creation requests fail, it still creates the project:

Screen Shot 2020-05-20 at 3 19 19 PM

Resulting teams created:

Screen Shot 2020-05-20 at 3 32 07 PM

Modal style changes:

Screen Shot 2020-05-20 at 3 38 14 PM

### CODE REVIEW

File: components/automate-ui/src/app/pages/project/list/project-list.component.ts [hunk 1]
```diff
@@ -44,6 +45,12 @@ export class ProjectListComponent implements OnInit, OnDestroy {
       'RULES_APPLIED': 'Applied'
   };
 
+  readonly teamsToCreate = {
+    'project-owner': 'Project Owner',
```
Comment (bcmdarroch) [inline components/automate-ui/src/app/pages/project/list/project-list.component.ts:49]: these should probably all be plural

File: components/automate-ui/src/app/pages/project/list/project-list.component.ts [hunk 2]
```diff
@@ -82,11 +90,26 @@ export class ProjectListComponent implements OnInit, OnDestroy {
...
+
         this.closeCreateModal();
 
         // This is issued periodically from projects-filter.effects.ts; we do it now
         // so the user doesn't have to wait.
         this.store.dispatch(new LoadOptions());
+
+        if (createTeamsWasChecked) {
+          for (const [teamID, teamName] of Object.entries(this.teamsToCreate)) {
+            this.store.dispatch(new CreateTeam({
```
Comment (bcmdarroch) [inline components/automate-ui/src/app/pages/project/list/project-list.component.ts:106]: I think to keep this component dumber, we'd want to handle this in an ngrx effect

File: components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.scss [hunk 3]
```diff
@@ -27,6 +27,10 @@ chef-modal {
 
       .checkbox-margin {
         margin-bottom: 30px;
+
+        #add-policy-checkbox {
+          margin-bottom: 30px;
+        }
```
Comment (srenatus) [inline components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.scss:33]: I don't get it. What does the class thing do? Is this added (to 60px), effectively?

File: components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.ts [hunk 4]
```diff
@@ -94,6 +97,11 @@ export class CreateObjectModalComponent implements OnInit, OnDestroy, OnChanges
     this.createForm.controls.addPolicies.setValue(this.addPolicies);
   }
 
+  updateTeamCheckbox(event: boolean): void {
```
Comment (srenatus) [inline components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.ts:100]: [nit] `event` seems a bit oddly named, perhaps? If it's a boolean, I'd go with `checked`. But I might be missing something.

File: components/automate-ui/src/app/pages/project/list/project-list.component.ts [hunk 5]
```diff
@@ -82,11 +90,26 @@ export class ProjectListComponent implements OnInit, OnDestroy {
       }))
       .subscribe(() => {
         this.creatingProject = false;
+
+        const createTeamsWasChecked = this.createProjectForm.controls['addTeams'].value;
```
Comment (srenatus) [inline components/automate-ui/src/app/pages/project/list/project-list.component.ts:94]: I don't know if and where we've settled on wrt. `this.createProjectForm.controls['addTeams'].value` vs `this.createProjectForm.controls.addTeams.value`

File: components/automate-ui/src/app/pages/project/list/project-list.component.ts [hunk 6]
```diff
@@ -82,11 +90,26 @@ export class ProjectListComponent implements OnInit, OnDestroy {
       }))
```
Comment (srenatus) [inline components/automate-ui/src/app/pages/project/list/project-list.component.ts:90]: [nit] ☝️ 
```typescript
filter(state => this.creatingProject && state === EntityStatus.loadingSuccess)
```

File: components/automate-ui/src/app/pages/project/list/project-list.component.ts [hunk 7]
```diff
@@ -55,7 +62,8 @@ export class ProjectListComponent implements OnInit, OnDestroy {
       name: ['', Validators.required],
       id: ['',
         [Validators.required, Validators.pattern(Regex.patterns.ID), Validators.maxLength(48)]],
-      addPolicies: [true]
+      addPolicies: [true],
```
Comment (srenatus) [inline components/automate-ui/src/app/pages/project/list/project-list.component.ts:65]: ❓ Does this not imply a default of true? I feel like we've stated that twice (with `addTeams = true` above)

File: components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.ts [hunk 8]
```diff
@@ -37,6 +37,7 @@ export class CreateObjectModalComponent implements OnInit, OnDestroy, OnChanges
   public modifyID = false; // Whether the edit ID form is open or not.
   public conflictError = false;
   public addPolicies = true;
+  public addTeams = true;
```
Comment (srenatus) [inline components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.ts:40]: Hmm is `create-object-modal` still the right name? It seems quite specific to have policy- and team-creation-related code in here 🤔

File: components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.ts (hunk 4 above)
Comment (tylercloke) [reply components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.ts:100]: I agree, once again following the policies pattern. I'll update em both.

File: components/automate-ui/src/app/pages/project/list/project-list.component.ts (hunk 7 above)
Comment (tylercloke) [reply components/automate-ui/src/app/pages/project/list/project-list.component.ts:65]: Yeah, it's a bit weird. This is the constructor for the FormBuilder which exists outside of the generic add modal component and is passed in. `addTeams` in that component is just a boolean we keep in sync with the checkbox state that we then use to drive the form. The first thing that component does is change these form values to true so defaulting them here does feel pointless.

File: components/automate-ui/src/app/pages/project/list/project-list.component.ts (hunk 5 above)
Comment (tylercloke) [reply components/automate-ui/src/app/pages/project/list/project-list.component.ts:94]: Yeah no idea which is considered more idiomatic. Would be a good #ui-team question.

File: components/automate-ui/src/app/pages/project/list/project-list.component.ts (hunk 2 above)
Comment (tylercloke) [reply components/automate-ui/src/app/pages/project/list/project-list.component.ts:106]: Hmmm I could see going either way on this one. Wanting to create a few teams as a result of how the user used the form feels like it makes sense to trigger in here. It feels weird that creating teams would be something the projects store would do.

On the other hand, I know we wanna keep these components as simple as possible within reason.

File: components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.scss (hunk 3 above)
Comment (tylercloke) [reply components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.scss:33]: Checkbox margin contains two checkboxes. This is adding a margin below the first checkbox whereas the checkbox margin is adding a margin below the container relative to the rest of the modal.

File: components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.spec.ts [hunk 9]
```diff
@@ -112,18 +112,24 @@ describe('CreateObjectModalComponent', () => {
     expect(store.dispatch).toHaveBeenCalledWith(new GetPolicies());
   });
 
-  it('upon opening, "also create policies" checkbox is checked', () => {
+  it('upon opening, "also create" checkboxs are checked', () => {
```
Comment (msorens) [inline components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.spec.ts:115]:
```suggestion
  it('upon opening, "also create" checkboxes are checked', () => {
```

File: components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.ts [hunk 10]
```diff
@@ -76,7 +77,9 @@ export class CreateObjectModalComponent implements OnInit, OnDestroy, OnChanges
       this.policiesUpdatedEvent.emit();
 
       if (this.createProjectModal) {
```
Comment (msorens) [inline components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.ts:79]: Regarding special casing in this code, here's a thought that makes it slightly less of a violation of the Open/Closed Principle: don't check who the caller is, rather just let the caller specify a need:
```suggestion
      if (this.associateTeamsAndPolicies) {
```

File: components/automate-ui/src/app/pages/project/list/project-list.component.ts [hunk 11]
```diff
@@ -77,16 +85,29 @@ export class ProjectListComponent implements OnInit, OnDestroy {
...
         // This is issued periodically from projects-filter.effects.ts; we do it now
         // so the user doesn't have to wait.
         this.store.dispatch(new LoadOptions());
+
+        if (createTeamsWasChecked) {
+          for (const [teamID, teamName] of Object.entries(this.teamsToCreate)) {
+            this.store.dispatch(new CreateTeam({
+              id: `${projectID}-${teamID}`,
+              name: `${projectName} ${teamName}`,
+              projects: [projectID]
```
Comment (msorens) [inline components/automate-ui/src/app/pages/project/list/project-list.component.ts:107]: Is there a reason we cannot make the addition of teams work more similar to the addition of policies? Seems odd that the two return different types of notifications, as shown.

If that approach is not viable, though, a comment on the current implementation:
Would like to have the notifications include the team id, so we do not see the repetitive "Created a new team". That is, should say "Created team "p1-foo-project-owners", etc.

image

File: components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.ts (hunk 8 above)
Comment (srenatus) [reply components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.ts:40]: `s/pattern we've established/hack/`  😉

Comment (srenatus) [reply components/automate-ui/src/app/components/create-object-modal/create-object-modal.component.ts:40]: Anyways, it's fine! 😃

**Pull request merged**